### DISCUSSION TITLE
Is there any option to put new independent tab on top instead of bottom?

### DISCUSSION THREAD

**Discussion by you74674** [piroor/treestyletab#3009] [Q&A] [answered]
There are options to control the behavior of new inserted child tab, like "the top of the tree" or "the end of the tree",
I want all new independent tabs to be on the top of the "root tree". 
Is there any way to achieve this? It seems like there's no option about this.

**Comment by piroor** [comment, marked as answer]
No such option for now, and I have no plan to implement it as a built-in feature, because it looks beyond TST's project scope. I think it can be done by something other extension (I don't know any existing implementation, and it may be developed as a new helper addon for TST) and I'm welcome to update TST compatible with such an extension.

**Reply by you74674** [reply]
Thank you for the reply.
Just curious, but I think it's quite natural to expect a tree having same behavior in every level(including the root level)?
Or the implementation is different between the root level and any other level?

For now the closest behavior is to create a new tab from pinned tab, so I think it should be possible to do in an extension.
Hope it won't be too difficult for me...

**Reply by piroor** [reply]
In short: I can implement the ability but I decided that I don't do it to keep the project scope.

* "Insertion position of new child tab" option is applied for a tab dropped onto an existing tab also. In other words, TST just applies this rule to any new tabs attached to the current tab, and doesn't control general insertion position of tabs opened from the current tab.
* Firefox doesn't allow us to put regular unpinned tabs between pinned tabs, so TST cannot attach child tabs to pinned tabs directly. Thus TST does a workaround: opens a group tab, set it as an alias of the pinned parent tab, and attach children to the alias tab. Because it looks natural that alias tab and children are placed near the real parent pinned tab, TST has ability to control insertion position of children opened from pinned tabs.

Both cases are related to tree management. On the other hand, insertion position of a new root tab looks not related to tree feature. I had to put the border somewhere to prevent TST become too bloated from too many features, and I  decided to put the border here.

**Reply by johnbwang** [reply]
I put up an add-on that puts new root tabs first (https://addons.mozilla.org/en-US/firefox/addon/tst-new-tabs-first/ and https://github.com/kytmizuno/tst-new-tabs-first). I trigger the move on browser.tabs.onCreated, expecting TST to have processed the tabs already. However, this means that correct behavior requires add-ons to be loaded in the correct order. When Firefox launches, TST may be loaded after tst-new-tabs-first, which break tst-new-tabs-first.

Is there a better event to listen to? Is there a TST event that I can listen to?

**Reply by you74674** [reply]
This is really good, thank you.
I just tried it and so far it works great.
About the loading order, I read some code from piroor.
https://github.com/piroor/tst-more-tree-commands/blob/trunk/background/background.js
Seems like `registerToTST` and `onMessageExternal.addListener` are important?

**Reply by piroor** [reply]
The order of addons handling `tabs.onCreated` looks uncontrollable. The registration logic just affects to notifications sent from TST.

I think it is the time to add a new notification type like `new-tab-processed`. I've experimentally added it with the commit 57a4a697e. The notified message has two properties: `tab` (the newly opened tab) and `ancestors` (parent tabs). It means "a new root level tab is opened" if `ancestors` is blank.

**Reply by johnbwang** [reply]
@piroor Thank you!

I've updated my extension's source to use new-tab-processed. Will update my extension in the Firefox store after TST is updated in the Firefox store.

**Reply by piroor** [reply]
I've added the document: https://github.com/piroor/treestyletab/wiki/API-for-other-addons#when-the-tst-finishes-its-handling-for-a-newly-opened-tab
Finally I've removed the `ancestorTabs` property. You'll need to update your code like as `s/ancestorTabs.length/tab.ancestorTabId.length/`,

**Reply by johnbwang** [reply]
Awesome! My add-on is now updated. @you74674, update both TST and new-tabs-first to fix the startup issue.

**Reply by you74674** [reply]
Thank you!
When there are too many tabs opened, it seems like the order of the process can still go wrong. When I try to restore a closed tab(in a tree), it will be inserted to the correct position first, and then moved to the top even though it should be in a tree.

**Reply by johnbwang** [reply]
@you74674 Okay, I'll take a look. We should continue the conversation at https://github.com/kytmizuno/tst-new-tabs-first/issues/1 since it's not directly related to TST.

**Reply by cstrahan** [reply]
@piroor  `message.tab.ancestorTabIds.length == 0` when opening a tab through the TST context menu, and also when opening a new tab through the "Open a new tab: Child Tab" key binding, which means that using https://github.com/johnbwang/tst-new-tabs-first/ results in new tabs (in either scenario) being moved as new roots in the tree, whereas I expect them to be children/siblings to the current tab (depending on TST settings).

Is there a way for third-party plugins to detect either scenario?

**Reply by cstrahan** [reply]
Some other details: https://github.com/piroor/treestyletab/issues/2543#issuecomment-1649723184

**Comment by ELLIOTTCABLE** [comment]
I took a gander around for other extensions that implement this, but I had no luck.

There's this one, but the advertised functionality doesn't currently work: [Tab Open/Close Control by F. Kolbe](https://addons.mozilla.org/en-US/firefox/addon/tab-open-close-control/).

I dived in to attempt to solve this with *CSS*, but ended up getting a little side-tracked. CSS, however, seems to be the "correct" solution: if you open new tabs at the *top*, then the *index* of all subsequent tabs changes every time you open a new root-level tab; you can't develop muscle-memory for, e.g., `:tab 26` within a particular session. If you maintain the tab-order, then reverse them on-*display*, at least *some* tab-indexes will remain static throughout a session …

It's not perfect, but my current, hacky solution is to reverse the entire `ul.tabs` using flex-box:

```css
/* Invert order of entire tab-tree.
 * Note: *Also reverses sub-trees*. */
#tabbar #all-tabs > ul.tabs {
  display: flex;
  flex-direction: column-reverse;
}
```

When I have more time, or get annoyed enough, I may attempt to improve on this such that *subtrees* are ordered in chronologically-increasing order, but the *root* tree is ordered in chronologically-decreasing order.

<img width="274" alt="Screen Shot 2021-11-17 at 13 44 58" src="https://user-images.githubusercontent.com/200/142271637-cde709d1-b834-4923-b306-b6dae6af82d4.png">

**Reply by you74674** [reply]
This is an interesting idea, thank you for sharing this!
